pyrate/core/phase_closure/closure_check.py

```diff
@@ -0,0 +1,93 @@
+from collections import namedtuple
+from pathlib import Path
+from typing import List, Dict
+import numpy as np
+from pyrate.core.shared import Ifg, dem_or_ifg
+from pyrate.core import config as cf
+from pyrate.core.shared import InputTypes
+from pyrate.core.phase_closure.mst_closure import find_signed_closed_loops
+from pyrate.core.phase_closure.sum_closure import sum_phase_values_for_each_loop
+from pyrate.core.phase_closure.plot_closure import plot_closure
+from pyrate.configuration import MultiplePaths
+from pyrate.core.logger import pyratelogger as log
+
+LARGE_DEVIATION_THRESHOLD_FOR_PIXEL = np.pi/4  # pi
+THRESHOLD_TO_REMOVE_IFG = 0.07  # ifgs with more than this fraction of pixels with error will be dropped
+LOOP_COUNT_FOR_THRESHOLD_TO_REMOVE_IFG = 2  # pixel with phase unwrap error in at least this many loops
+PHASE_UNWRAP_ERROR_THRESHOLD = 5  # pixel with phase unwrap error in more than this many ifgs will be flagged
+MAX_LOOP_LENGTH = 4  # loops upto this many edges are considered for closure checks
+SUBTRACT_MEDIAN_IN_CLOSURE_CHECK = True
+
+
+def detect_ps_with_unwrapping_errors(check_ps, num_occurences_each_ifg):
+    nrows, ncols, n_ifgs = check_ps.shape
+    ps_unwrap_error = np.zeros(shape=(nrows, ncols), dtype=np.int16)
+    for i in range(n_ifgs):
+        ps_idx = check_ps[:, :, i] == num_occurences_each_ifg[i]
+        ps_unwrap_error[ps_idx] += 1  # number of IFGs with unwrapping errors per PS
+    # PS pixels with unwrapping errors in one or more SBAS IFGs will be marked.
+    # mark_ix = ps_unwrap_error > 0  # don't need to output this
+
+    # keep_ix = ~ (ps_unwrap_error >= PHASE_UNWRAP_ERROR_THRESHOLD)
+    # log.info(f'Of {nrows * ncols} pixels, {np.sum(~keep_ix)} '
+    #          f'have phase unwrapping error in {PHASE_UNWRAP_ERROR_THRESHOLD} or more pixels')
+    # can move mark_ix an keep_ix in wrapper if at all required
```

Comment:
To be discussed what to do with pixels with unwrapping errors in multiple closure loops. We could think about masking these pixels out, similar as is done for the coherence masking.

Reply:
Hi @tfuhrmann, concerned about this comment:

` I think something is wrong with the plots (and/or the creation of non-redundant loops). `

Can you elaborate? Are there any loop that's repeated? Are these invalid loops?

Reply:
> Hi @tfuhrmann, concerned about this comment:
> 
> `I think something is wrong with the plots (and/or the creation of non-redundant loops).`
> 
> Can you elaborate? Are there any loop that's repeated? Are these invalid loops?

In the image posted above, the second column contains the same images (i.e. the same loop). I've printed the loops used to calculate the closure check and I can't spot the duplicates there (see below Python output for the `loop` variable in `sum_closure.py`. So, I think this duplication is only related to the plotting. 
Furthermore, I was a bit confused by the number of loops (73 from 30 interferograms). From my experience with the Matlab version of the code, the number of loops is usually smaller than the number of interferograms. Note that the large number of loops would also explain the long runtime for a data set containing more epochs and interferograms. To further investigate, I've replicated the loop calculation for the crop A data set in the Matlab version. Running `sbas_mst_closure.m` with that results in the following 18 loops (epoch denotes the image dates: e.g. 1=20180106, 13=20180717; ifg denotes the interferograms: e.g. 1=20180106-20180130, including the sign used to close the loops):
![matlab_code_loops_cropA](https://user-images.githubusercontent.com/20388928/102147689-414a5c00-3ebf-11eb-91a1-3f099acc2a33.PNG)
As you can see most of the loops (except for two) consist of three interferograms/edges only.

For comparison, loops calculated by the PyRate closure loop algorithm for the Mexico Crop A test data set:
```
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 7, 17)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 7, 17)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 23)), sign=-1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 7, 17)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 7, 17)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 30)), sign=-1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 7, 17)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 7, 17)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=-1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 7, 17)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 7, 17)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 7, 17)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 7, 17)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 4, 12)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 7, 17)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 7, 17)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 7, 17)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 7, 17)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 30)), sign=-1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=-1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 4, 12)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 30)), sign=-1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=-1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 19)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 11)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 6, 11)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 30)), sign=-1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 11)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 6, 11)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 11)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 6, 11)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 31)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 6, 11)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 6, 11)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 19)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=-1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 4, 12)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=-1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 19)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 19)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 4, 12)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 5, 6), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 3, 19)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 6)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 18)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 6)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 6)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 6)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 6)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 30), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 30), second=datetime.date(2018, 3, 7)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 6)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 3, 19)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 6)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 6)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 19)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 6)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 6)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 19)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 6)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 18)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 18)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 30), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 1, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 5, 18)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 5, 18)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 4, 12), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 3, 19)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 18)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 18)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 18)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 4, 12)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 5, 18)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 18)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 3, 19)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 5, 18)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 19)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 18)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 18)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 3, 19)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 5, 18)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 1, 30), second=datetime.date(2018, 4, 12)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 30), second=datetime.date(2018, 3, 7)), sign=-1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 1, 30), second=datetime.date(2018, 4, 12)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 1, 30)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 1, 30), second=datetime.date(2018, 3, 7)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 19)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 3, 19)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 1, 30)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 30)), sign=-1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 6, 23)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 6, 23)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 19)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 31)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 19)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 31)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 5, 30)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 31)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 31), second=datetime.date(2018, 4, 12)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 4, 12)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 1, 6), second=datetime.date(2018, 3, 19)), sign=1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 3, 31)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 19)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 31)), sign=1)]
[SignedEdge(edge=Edge(first=datetime.date(2018, 3, 19), second=datetime.date(2018, 5, 30)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 3, 19)), sign=-1), SignedEdge(edge=Edge(first=datetime.date(2018, 3, 7), second=datetime.date(2018, 5, 30)), sign=1)]
```